cluster-api/pkg/apis/cluster/v1alpha1/machinedeployment_types.go

```diff
@@ -0,0 +1,288 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"log"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/endpoints/request"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	metav1validation "k8s.io/apimachinery/pkg/apis/meta/v1/validation"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	utilvalidation "k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	"k8s.io/kube-deploy/cluster-api/pkg/apis/cluster"
+	"k8s.io/kube-deploy/cluster-api/pkg/apis/cluster/common"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineDeployment
+// +k8s:openapi-gen=true
+// +resource:path=machinedeployments,strategy=MachineDeploymentValidationStrategy
+type MachineDeployment struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   MachineDeploymentSpec   `json:"spec,omitempty"`
+	Status MachineDeploymentStatus `json:"status,omitempty"`
+}
+
+// MachineDeploymentSpec defines the desired state of MachineDeployment
+type MachineDeploymentSpec struct {
+	// Number of desired machines. Defaults to 1.
+	// This is a pointer to distinguish between explicit zero and not specified.
+	Replicas *int32 `json:"replicas,omitempty"`
+
+	// Label selector for machines. Existing MachineSets whose machines are
```

Comment:
For replica sets, it seems there is an assumption that there aren't sets with overlapping selectors, otherwise things get tricky as per link below. Is this the same assumption here?

https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/#labels-on-a-replicaset

Reply:
Yes, same assumption. Deployments use the concept of "pod-template-hash", we can probably do something similar.
https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#pod-template-hash-label